From df54785a8d0e9f5501adce617c9c1438980daae7 Mon Sep 17 00:00:00 2001
From: Liu Shixin <liushixin2@huawei.com>
Date: Thu, 30 Jul 2026 10:25:33 +0800
Subject: [PATCH 958/960] downstream: iommu/ummu-core: support for ummu-core
 dkms

Add the macros, structures and function declarations required by dkms.

Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 include/linux/ummu_core.h      | 452 ++++++++++++++++++++++++++++++++-
 include/uapi/linux/ummu_core.h | 144 +++++++++++
 2 files changed, 592 insertions(+), 4 deletions(-)
 create mode 100644 include/uapi/linux/ummu_core.h

diff --git a/include/linux/ummu_core.h b/include/linux/ummu_core.h
index 74ed202d1d89..d91b1714607b 100644
--- a/include/linux/ummu_core.h
+++ b/include/linux/ummu_core.h
@@ -7,6 +7,7 @@
 #ifndef _UMMU_CORE_H_
 #define _UMMU_CORE_H_
 
+#include <uapi/linux/ummu_core.h>
 #include <linux/iommu.h>
 #include <linux/uuid.h>
 #include <linux/xarray.h>
@@ -14,6 +15,16 @@
 
 #define eid_t u128
 
+#define UB_MAX_TID_BITS 20U
+#define UB_MAX_TID ((1 << UB_MAX_TID_BITS) - 1)
+
+#define UMMU_NO_TID 0U
+#define UMMU_INVALID_TID UB_MAX_TID
+
+#define UMMU_DEV_WRITE 1
+#define UMMU_DEV_READ 2
+#define UMMU_DEV_ATOMIC 4
+
 enum eid_type {
 	EID_NONE = 0,
 	EID_BYPASS,
@@ -26,10 +37,24 @@ enum tid_alloc_mode {
 	TID_ALLOC_NORMAL = 2,
 };
 
-enum ummu_mapt_mode {
-	MAPT_MODE_TABLE = 0,
-	MAPT_MODE_ENTRY,
-	MAPT_MODE_END,
+enum ummu_resource_type {
+	UMMU_BLOCK,
+	UMMU_QUEUE,
+	UMMU_QUEUE_LIST,
+	UMMU_CNT,
+	UMMU_TID_RES,
+};
+
+enum default_tid_ops_types {
+	PASID_OPS,
+	DEFAULT_OPS,
+	TID_OPS_MAX,
+};
+
+enum ummu_register_type {
+	REGISTER_TYPE_GLOBAL,
+	REGISTER_TYPE_NORMAL,
+	REGISTER_TYPE_MAX,
 };
 
 struct iova_slot;
@@ -37,6 +62,42 @@ struct ummu_tid_manager;
 struct ummu_base_domain;
 struct ummu_core_device;
 
+struct block_args {
+	u32 index;
+	int block_size_order;
+	phys_addr_t out_addr;
+};
+
+struct queue_args {
+	phys_addr_t pcmdq_base;
+	phys_addr_t pcplq_base;
+	phys_addr_t ctrl_page;
+};
+
+struct tid_args {
+	u8 pcmdq_order;
+	u8 pcplq_order;
+	size_t blk_exp_size;
+	u64 hw_cap;
+};
+
+struct resource_args {
+	enum ummu_resource_type type;
+	union {
+		struct block_args block;
+		struct queue_args queue;
+		struct queue_args *queues;
+		struct tid_args tid_res;
+		u32 ummu_cnt;
+		u32 block_index;
+	};
+	int align;
+};
+
+struct ummu_param {
+	enum ummu_mapt_mode mode;
+};
+
 struct ummu_tid_param {
 	struct device *device;
 	enum ummu_mapt_mode mode;
@@ -53,15 +114,25 @@ struct tdev_attr {
 
 /**
  * struct ummu_core_ops - ummu ops for normal use, expand from iommu_ops.
+ * @get_resource: Get resource for SVA.
+ * @put_resource: Put resource for SVA.
  * @add_eid: Add EID to the UMMU device.
  * @del_eid: Add EID to the UMMU device.
+ * @invalidate_cfg: invalid configuration table by tid.
+ * @cfg_syn_all: synchronize all configuration table.
+ * @cfg_syn: synchronize configuration table by tid.
  * @tdev_support_attr: Check whether the UMMU device supports the tdev attribute.
  */
 struct ummu_core_ops {
+	int (*get_resource)(struct ummu_base_domain *d, struct resource_args *arg);
+	void (*put_resource)(struct ummu_base_domain *d, struct resource_args *arg);
 	int (*add_eid)(struct ummu_core_device *dev, guid_t *guid, eid_t eid,
 		       enum eid_type type);
 	void (*del_eid)(struct ummu_core_device *dev, guid_t *guid, eid_t eid,
 			enum eid_type type);
+	int (*invalidate_cfg)(struct ummu_base_domain *d);
+	void (*cfg_sync_all)(struct ummu_base_domain *d);
+	void (*cfg_sync)(struct ummu_base_domain *d);
 	bool (*tdev_support_attr)(struct ummu_core_device *dev, struct tdev_attr *attr);
 };
 
@@ -105,6 +176,44 @@ struct ummu_tid_manager {
 	void *tid_data;
 };
 
+struct ummu_core_tid_args {
+	const struct tid_ops *tid_ops;
+	u32 max_tid;
+	u32 min_tid;
+};
+
+struct ummu_core_init_args {
+	const struct ummu_core_ops *core_ops;
+	struct ummu_core_tid_args tid_args;
+	const struct iommu_ops *iommu_ops;
+	struct device *hwdev;
+};
+
+/* Memory traffic monitoring of the UB device */
+struct ummu_mpam {
+#define UMMU_DEV_SET_MPAM	(1 << 0)
+#define UMMU_DEV_GET_MPAM	(1 << 1)
+#define UMMU_DEV_SET_USER_MPAM_EN	(1 << 2)
+#define UMMU_DEV_GET_USER_MPAM_EN	(1 << 3)
+	int flags;
+	eid_t eid;
+	int tid;
+	int partid;
+	int pmg;
+	int s1mpam;
+	int user_mpam_en;
+};
+
+enum ummu_device_config_type {
+	UMMU_MPAM = 0,
+};
+
+#if IS_ENABLED(CONFIG_UB_UMMU_CORE_DRIVER)
+extern const struct tid_ops *ummu_core_tid_ops[TID_OPS_MAX];
+#else
+static const struct tid_ops *ummu_core_tid_ops[TID_OPS_MAX];
+#endif /* CONFIG_UB_UMMU_CORE_DRIVER */
+
 static inline struct ummu_core_device *to_ummu_core(struct iommu_device *iommu)
 {
 	return container_of(iommu, struct ummu_core_device, iommu);
@@ -116,6 +225,14 @@ to_ummu_base_domain(struct iommu_domain *dom)
 	return container_of(dom, struct ummu_base_domain, domain);
 }
 
+static inline void tdev_attr_init(struct tdev_attr *attr)
+{
+	attr->dma_attr = DEV_DMA_COHERENT;
+	attr->name = NULL;
+	attr->priv = NULL;
+	attr->priv_len = 0;
+}
+
 #ifdef CONFIG_UB_UMMU_CORE
 /* EID API */
 /**
@@ -211,4 +328,331 @@ static inline int ummu_drain_pages(struct iova_slot *slot, dma_addr_t iova,
 	return -EOPNOTSUPP;
 }
 #endif /* CONFIG_UB_UMMU_CORE */
+
+#if IS_ENABLED(CONFIG_UB_UMMU_CORE_DRIVER)
+/* UMMU SVA API */
+/**
+ * Grant va range permission to sva.
+ * @sva: related sva handle.
+ * @va: va start
+ * @size: va size
+ * @perm: permission
+ * @cookie: struct ummu_token_info*
+ *		if (!cookie) {
+ *			do not use cookie check.
+ *		} else if (cookie->input == 0) {
+ *			use this cookie->tokenval
+ *		} else if (cookie->input == 1) {
+ *			cookie->tokenval = generate new one
+ *		} else {
+ *			invalid para
+ *		}
+ *
+ * Return: 0 on success, or an error.
+ */
+int ummu_sva_grant_range(struct iommu_sva *sva, void *va, size_t size, int perm,
+			 void *cookie);
+
+/**
+ * Ungrant va range permission from sva.
+ * @sva: related sva handle.
+ * @va: va start
+ * @size: va size
+ * @cookie: va related cookie,struct ummu_token_info*
+ *		if (!cookie) {
+ *			do not use cookie check.
+ *		} else {
+ *			ungrant by cookie->tokenval
+ *		}
+ *
+ * Return: 0 an success, or an error.
+ */
+int ummu_sva_ungrant_range(struct iommu_sva *sva, void *va, size_t size,
+			   void *cookie);
+
+/**
+ * Get tid from dev or sva.
+ * @dev: related device.
+ * @sva: if sva is set, return sva mode related tid; otherwise
+ *	 return the dma mode tid.
+ * @tidp: tid returned here.
+ *
+ * Return: 0 on success, or an error.
+ */
+int ummu_get_tid(struct device *dev, struct iommu_sva *sva, u32 *tidp);
+
+/**
+ * Get iommu_domain by tid and dev.
+ * @dev: related device.
+ * @tid: tid
+ *
+ * Return: iommu_domain or NULL if failed.
+ */
+struct iommu_domain *ummu_core_get_domain_by_tid(struct device *dev,
+						 u32 tid);
+
+/**
+ * Check whether the UMMU works in ksva mode.
+ * @domain: related iommu domain
+ *
+ * Return: true or false.
+ */
+bool ummu_is_ksva(struct iommu_domain *domain);
+
+/**
+ * Check whether the UMMU works in sva mode.
+ * @domain: related iommu domain
+ *
+ * Return: true or false.
+ */
+bool ummu_is_sva(struct iommu_domain *domain);
+
+/**
+ * Bind device to a process mm.
+ * @dev: related device.
+ * @mm: process memory management.
+ * @drvdata: ummu_param related to tid.
+ *		if (!drvdata) {
+ *			sva is in the bypass mapt mode.
+ *		} else {
+ *			follow the drvdata->mode to set mapt mode.
+ *		}
+ *
+ * Return: sva handle or NULL if failed.
+ */
+struct iommu_sva *ummu_sva_bind_device(struct device *dev, struct mm_struct *mm,
+				       struct ummu_param *drvdata);
+
+/**
+ * Bind device to kernel mm.
+ * @dev: related device.
+ * @drvdata: ummu_param related to tid. ksva doesn't support bypass mapt.
+ *
+ * Return: sva handle or NULL if failed.
+ */
+struct iommu_sva *ummu_ksva_bind_device(struct device *dev,
+					struct ummu_param *drvdata);
+void ummu_sva_unbind_device(struct iommu_sva *handle);
+void ummu_ksva_unbind_device(struct iommu_sva *handle);
+
+/* UMMU CORE API */
+/**
+ * Initialiase ummu core device.
+ * @ummu_core: ummu core device.
+ * @args: ummu core init args.
+ * UMMU driver should carefully choose the args based on its requirement.
+ *	iommu_ops is mandatory.
+ *	a. the ummu device need tid allocation capability.
+ *		a.1 default tid strategies satisfy the ummu device
+ *			-> set tid_ops form ummu_core_tid_ops[TID_OPS_MAX]
+ *		a.2 default tid strategies do not satisfy the ummu device
+ *			-> implement a new tid_ops in the driver.
+ *	b. the ummu device need ummu core ops capability.
+ *		-> set core_ops.
+ *	c. the ummu device has related hwdev.
+ *		-> set hwdev.
+ */
+int ummu_core_device_init(struct ummu_core_device *ummu_core,
+			  struct ummu_core_init_args *args);
+/**
+ * Deinitialiase ummu core device.
+ * @ummu_core: ummu core device.
+ */
+void ummu_core_device_deinit(struct ummu_core_device *ummu_core);
+
+/**
+ * Register ummu core device to the ummu framework.
+ * @ummu_core: ummu core device.
+ * @type: register type.
+	REGISTER_TYPE_GLOBAL: register the ummu device as the global device,
+		The ummu device will be the device handle all request.
+		e.g. 1. add_eid/del_eid 2. provide ubus iommu ops. etc.
+
+	REGISTER_TYPE_NORMAL: follow the iommu_device register. will not be
+		related to the global device. it work as a normal iommu device.
+ */
+int ummu_core_device_register(struct ummu_core_device *ummu_core,
+			      enum ummu_register_type type);
+/**
+ * Unregister ummu core device from the ummu framework.
+ * @dev: the ummu_core device tid belongs to.
+ */
+void ummu_core_device_unregister(struct ummu_core_device *dev);
+
+/**
+ * Invalidate ummu global configuration by tid.
+ * @tid: tid
+ * Return: 0 on success, or an error.
+ */
+int ummu_core_invalidate_cfg_table(u32 tid);
+
+/* UMMU TID API */
+/**
+ * Alloc a tid from ummu framework, and alloc related pasid.
+ * @dev: the allocated tid will be attached to.
+ * @drvdata: ummu_tid_param related to tid
+ * @tidp: the allocated tid returned here.
+ *
+ * Return: 0 on success, or an error.
+ */
+int ummu_core_alloc_tid(struct ummu_core_device *dev,
+			struct ummu_tid_param *drvdata, u32 *tidp);
+
+/**
+ * Free a tid to ummu framework.
+ * @dev: the ummu_core device tid belongs to.
+ * @tid: token id.
+ */
+void ummu_core_free_tid(struct ummu_core_device *dev, u32 tid);
+
+/**
+ * Get mapt_mode related to the tid.
+ * @dev: the ummu_core device tid belongs to.
+ * @tid: token id.
+ *
+ * Return: ummu_mapt_mode, negative for an error.
+ */
+enum ummu_mapt_mode ummu_core_get_mapt_mode(struct ummu_core_device *dev,
+					    u32 tid);
+
+/**
+ * Get device related to the tid.
+ * It will increase the ref count of the device.
+ * @dev: the ummu_core device tid belongs to.
+ * @tid: token id.
+ *
+ * Return: device or NULL if failed.
+ */
+struct device *ummu_core_get_device(struct ummu_core_device *dev, u32 tid);
+void ummu_core_put_device(struct device *dev);
+
+/**
+ *  Allocate a virtual device to hold a tid.
+ * @attr: attributes of tdev
+ * @ptid: tid pointer
+ * Return: device on success or NULL error.
+ */
+struct device *ummu_core_alloc_tdev(struct tdev_attr *attr, u32 *ptid);
+
+/**
+ * Free the virtual device
+ * @dev: Return value allocated by ummu_core_alloc_tdev
+ *
+ * Return: 0 on success or an error.
+ */
+int ummu_core_free_tdev(struct device *dev);
+#else
+static inline int ummu_sva_grant_range(struct iommu_sva *sva, void *va,
+				       size_t size, int perm, void *cookie)
+{
+	return -EOPNOTSUPP;
+}
+
+static inline int ummu_sva_ungrant_range(struct iommu_sva *sva, void *va,
+					 size_t size, void *cookie)
+{
+	return -EOPNOTSUPP;
+}
+
+static inline int ummu_get_tid(struct device *dev, struct iommu_sva *sva,
+			       u32 *tidp)
+{
+	return -EOPNOTSUPP;
+}
+
+static inline struct iommu_domain *
+ummu_core_get_domain_by_tid(struct device *dev, u32 tid)
+{
+	return NULL;
+}
+
+static inline bool ummu_is_ksva(struct iommu_domain *domain)
+{
+	return false;
+}
+
+static inline bool ummu_is_sva(struct iommu_domain *domain)
+{
+	return false;
+}
+
+static inline struct iommu_sva *ummu_sva_bind_device(struct device *dev,
+						     struct mm_struct *mm,
+						     struct ummu_param *drvdata)
+{
+	return NULL;
+}
+
+static inline struct iommu_sva *
+ummu_ksva_bind_device(struct device *dev, struct ummu_param *drvdata)
+{
+	return NULL;
+}
+
+static inline void ummu_sva_unbind_device(struct iommu_sva *handle)
+{
+}
+static inline void ummu_ksva_unbind_device(struct iommu_sva *handle)
+{
+}
+static inline int ummu_core_device_init(struct ummu_core_device *ummu_core,
+					struct ummu_core_init_args *args)
+{
+	return -EOPNOTSUPP;
+}
+
+static inline void ummu_core_device_deinit(struct ummu_core_device *ummu_core)
+{
+}
+static inline int ummu_core_device_register(struct ummu_core_device *ummu_core,
+					    enum ummu_register_type type)
+{
+	return -EOPNOTSUPP;
+}
+
+static inline void ummu_core_device_unregister(struct ummu_core_device *dev)
+{
+}
+static inline int ummu_core_invalidate_cfg_table(u32 tid)
+{
+	return -EOPNOTSUPP;
+}
+
+static inline int ummu_core_alloc_tid(struct ummu_core_device *dev,
+				      struct ummu_tid_param *drvdata,
+				      u32 *tidp)
+{
+	return -EOPNOTSUPP;
+}
+
+static inline void ummu_core_free_tid(struct ummu_core_device *dev,
+				      u32 tid)
+{
+}
+static inline enum ummu_mapt_mode
+ummu_core_get_mapt_mode(struct ummu_core_device *dev, u32 tid)
+{
+	return MAPT_MODE_END;
+}
+
+static inline struct device *ummu_core_get_device(struct ummu_core_device *dev,
+						  u32 tid)
+{
+	return NULL;
+}
+
+static inline void ummu_core_put_device(struct device *dev)
+{
+}
+
+static inline struct device *ummu_core_alloc_tdev(struct tdev_attr *attr, u32 *ptid)
+{
+	return NULL;
+}
+
+static inline int ummu_core_free_tdev(struct device *dev)
+{
+	return -EOPNOTSUPP;
+}
+#endif /* CONFIG_UB_UMMU_CORE_DRIVER */
 #endif /* _UMMU_CORE_H_ */
diff --git a/include/uapi/linux/ummu_core.h b/include/uapi/linux/ummu_core.h
new file mode 100644
index 000000000000..8ed0aba10b00
--- /dev/null
+++ b/include/uapi/linux/ummu_core.h
@@ -0,0 +1,144 @@
+/* SPDX-License-Identifier: GPL-2.0+ WITH Linux-syscall-note */
+/*
+ * Copyright (c) 2025 HiSilicon Technologies Co., Ltd. All rights reserved.
+ * Description: UMMU-CORE defines and function prototypes.
+ */
+
+#ifndef _UAPI_UMMU_CORE_H_
+#define _UAPI_UMMU_CORE_H_
+
+#include <linux/types.h>
+
+enum ummu_mapt_perm {
+	MAPT_PERM_W = 1,
+	MAPT_PERM_R = 2,
+	MAPT_PERM_RW = 3,
+	MAPT_PERM_ATOMIC_W = 5,
+	MAPT_PERM_ATOMIC_R = 6,
+	MAPT_PERM_ATOMIC_RW = 7
+};
+
+enum ummu_mapt_mode {
+	MAPT_MODE_TABLE = 0,
+	MAPT_MODE_ENTRY = 1,
+	MAPT_MODE_END
+};
+
+enum ummu_ebit_state {
+	UMMU_EBIT_OFF = 0,
+	UMMU_EBIT_ON = 1,
+	UMMU_EBIT_END
+};
+
+struct ummu_token_info {
+	/* 0:input, 1:output */
+	unsigned int input;
+	unsigned int tokenVal;
+
+	struct {
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+	} kabi_reserve;
+};
+
+struct ummu_tid_attr {
+	enum ummu_mapt_mode mode;
+
+	struct {
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+	} kabi_reserve;
+};
+
+struct ummu_seg_attr {
+	struct ummu_token_info *token;
+	enum ummu_ebit_state e_bit;
+
+	struct {
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+	} kabi_reserve;
+};
+
+struct ummu_tid_info {
+	unsigned char pcmdq_order;
+	unsigned char pcplq_order;
+	int expan;
+	unsigned long blk_exp_size;
+	unsigned long dev;
+	unsigned int tid;
+	unsigned int ummu_cnt;
+	enum ummu_mapt_mode mode;
+	__u64 hw_cap;
+	__u64 va;
+	__u64 size;
+
+	struct {
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+		__u64:64;
+	} kabi_reserve;
+};
+
+#define HW_CAP_EXPAN (1UL << 0)
+#define HW_CAP_DOUBLE_PLBI (1UL << 1)
+#define HW_CAP_KCMD_PLBI (1UL << 2)
+
+#define TID_MISC_NAME "ummu/tid"
+#define TID_DEVICE_NAME "/dev/" TID_MISC_NAME
+
+#define UMMU_IOC_MAGIC ((unsigned int)'k')
+
+#define UMMU_IOCALLOC_TID _IOW(UMMU_IOC_MAGIC, 4U, struct ummu_tid_info)
+#define UMMU_IOCFREE_TID _IOW(UMMU_IOC_MAGIC, 5U, struct ummu_tid_info)
+#define UMMU_IOCPLBI_VA _IOW(UMMU_IOC_MAGIC, 6U, struct ummu_tid_info)
+#define UMMU_IOCPLBI_ALL _IOW(UMMU_IOC_MAGIC, 7U, struct ummu_tid_info)
+
+#define BLOCK_SIZE_4K		0x00001000
+#define BLOCK_SIZE_16K		0x00004000
+#define BLOCK_SIZE_32K		0x00008000
+#define BLOCK_SIZE_64K		0x00010000
+#define BLOCK_SIZE_128K		0x00020000
+#define BLOCK_SIZE_2M		0x00200000
+
+#define MMAP_TYPE_BLOCK (1UL)
+#define MMAP_TYPE_MUTIQ (2UL)
+
+#define BLOCK_MAX_INDEX 512
+#define COMPRESSED_SHIFT 12
+#define UMMU_OFFSET_OP_SHIFT PAGE_SHIFT
+#define UMMU_OFFSET_INDEX_SHIFT (UMMU_OFFSET_OP_SHIFT + 2U)
+#define UMMU_OFFSET_SIZE_SHIFT (UMMU_OFFSET_INDEX_SHIFT + 10U)
+
+/*
+ *|20bit|10bit|2bit|page_shift|
+ *| tid |index|op  | rsv      |
+ */
+
+#define UMMU_OFFSET_BLK(index, tid) (((MMAP_TYPE_BLOCK) << (UMMU_OFFSET_OP_SHIFT)) \
+					| (((__u64)(index)) << (UMMU_OFFSET_INDEX_SHIFT)) \
+					| (((__u64)(tid)) << (UMMU_OFFSET_SIZE_SHIFT)))
+
+#define UMMU_OFFSET_QUE(tid) ((MMAP_TYPE_MUTIQ) << (UMMU_OFFSET_OP_SHIFT) \
+				| (((__u64)(tid)) << (UMMU_OFFSET_INDEX_SHIFT)))
+
+#endif /* UMMU_CORE_H */
-- 
2.43.0

